MAINT: Reduce allocation size of empty (0 size) arrays to 1 byte

mattip opened the pull request:
Restart #15788 from @eric-wieser.
- Allocate a minimum 1 byte when the data chunk should be 0 bytes, to avoid undefined behaviour
- Make allocation match the free calculation of how much memory to release. This closes #16410 and #20289

eric-wieser committed:
MAINT: Shrink the size of memory allocations to 1 bytes for arrays with `size == 0`.

In order to avoid undefined behavior, we need to ensure the generated strides are all zero.

mattip committed:
if any dims is 0, allocate array of length 1

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -825,11 +825,9 @@ PyArray_NewFromDescr_int(
-         * Note: always sync this with calls to PyDataMem_UserFREE
          */
         if (nbytes == 0) {
-            nbytes = descr->elsize ? descr->elsize : 1;
```
Can we ensure all the strides are set to zero to avoid any concerns of undefined behavior?

mattip replied:
OK, but `array_strides_set` will allow resetting the strides to any values.

takagi replied:
I have the following behavior when a 0-size array is reshaped in NumPy 1.23:
```
>>> import numpy as np
>>> a = np.ones((2, 0, 3))
>>> a.strides
(0, 0, 0)
>>> b = a.reshape((3, 0, 2))
>>> b.strides
(16, 16, 8)
```
Do you think `reshape` should be also fixed to return `b` with strides `(0, 0, 0)` here, @eric-wieser @mattip ? Thanks!

ref. https://github.com/cupy/cupy/pull/6807

mattip replied:
That does look like a bug. Please open a new issue for this, the PR here was merged and further discussion will get lost.

takagi replied:
Thanks! I opened #21918.

mattip committed:
fix from review: make sure all strides are 0 when nbytes is 0

seberg commented:
I am wondering if we should worry about limiting the total size also in this case.  But  I cannot think of any place where it would actually be used (i.e. we assume that a subset of axis cannot overflow when combined).

So, I would feel a bit safer to keep that check unmodified, but I am OK to not worry about it now.

mattip commented:
> this case, that check

which ones?

mattip committed:
fixes from review, fixes to assumptions about strides

mattip commented:
> Sorry, the current loop ignores 0s, so if you create an array of shape (0, 2**60, 2**60) it currently fails just like (2**60, 2**60, 0) fails.

I restored this behaviour.

There were a few places in the code that assumed `strides[-1] == elsize`, so I wonder if this change will break downstream code.

eric-wieser commented:
> There were a few places in the code that assumed `strides[-1] == elsize`, so I wonder if this change will break downstream code.

I'd hoped this was the type of thing that `RELAXED_STRIDES_CHECKING` was supposed to transition downstream code away from.

eric-wieser commented on numpy/core/src/multiarray/getset.c:
```diff
@@ -513,6 +513,7 @@ array_descr_set(PyArrayObject *self, PyObject *arg, void *NPY_UNUSED(ignored))
         /* resize on last axis only */
         int axis = PyArray_NDIM(self) - 1;
         if (PyArray_DIMS(self)[axis] != 1 &&
+                PyArray_STRIDES(self)[axis] != 0 &&
```
Maybe this should be
```suggestion
        if (PyArray_DIMS(self)[axis] > 1 &&
```

seberg replied:
This suggestion somewhat makes sense to me.  OTOH, I don't want anyone to set `arr.dtype` to begin with, and it would lead to non-zero strides.
So I think, I am fine with just forcing that one person who runs into it to doing weird stuff.  They should be doing different stuff anyway...

@eric-wieser I am planning to just merge thi sin a bit.  If you care and its merged, just make a note and I will follow-up.

mattip replied:
Since this makes sense let's put it in.

eric-wieser replied:
Isn't `view` still implemented via this code (making a copy of the array container then setting `.dtype`?)

seberg replied:
Yeah, you are right `view` does indeed use this.  The tests are failing here for some reason though, I didn't investigate it yet:
```
../../builds/venv/lib/python3.8/site-packages/numpy-1.23.0.dev0+1181.gadd7834a6-py3.8-linux-x86_64.egg/numpy/linalg/tests/test_linalg.py:320: in _make_strided_cases
    for a, a_label in _stride_comb_iter(case.a):
        a          = array([[0.36561909],
       [0.8016026 ],
       [0.78273559],
       [0.70135538],
       [0.62277659]])
        a_label    = 'stride_xxx_0'
        b          = array([0.49368265, 0.8405377 , 0.71209699, 0.44390898, 0.03103486])
        b_label    = 'stride_-4'
        case       = <LinalgCase: 0x4>
        new_case   = <LinalgCase: 5x1_stride_xxx_0_stride_-4>
        new_cases  = [<LinalgCase: single_stride_+1_+1_stride_+1>, <LinalgCase: single_stride_+1_+1_stride_+3>, <LinalgCase: single_stride_...stride_+1_+3_stride_+1>, <LinalgCase: single_stride_+1_+3_stride_+3>, <LinalgCase: single_stride_+1_+3_stride_-4>, ...]
../../builds/venv/lib/python3.8/site-packages/numpy-1.23.0.dev0+1181.gadd7834a6-py3.8-linux-x86_64.egg/numpy/linalg/tests/test_linalg.py:291: in _stride_comb_iter
    xi.view(np.uint32).fill(0xdeadbeef)
E   ValueError: To change to a dtype of a different size, the last axis must be contiguous
        new_shape  = [0, 4]
        repeats    = (1, 1)
        slices     = (slice(None, None, 1), slice(None, None, 1))
        stride_set = [(1, 3, -4), (1, 3, -4)]
        x          = array([], shape=(0, 4), dtype=float64)
        xi         = array([], shape=(0, 4), dtype=float64)
```

mattip replied:
Reverted.

eric-wieser replied:
I think it would be worth understanding the failure. The `PyArray_DIMS(self)[axis] != 1 && PyArray_STRIDES(self)[axis] != 0` logic feels wrong to me, as it means zero-size arrays may or may not reach this error depending on their stride (which can be anything as long as their allocation is ok).

seberg replied:
The reason is that we are inspecting the stride of a `(0, 4)` array.  We are setting the strides to `(0, 0)`, so the second dimension is "not contiguous" (not `stride == itemsize length <= 1`).
The fix might be to do a check for whether the array itself is contiguous (which captures the empty part).

seberg replied:
But, agruably, the current code is no better, since it would allow overriding a stride 0 dimension and giving it a non-zero stride, which could access out-of-bounds memory:
```
arr = np.ones(1, dtype="i8")
arr = np.broadcast_to(arr, 10)
arr.dtype = "i1"
print(arr)  # oops
```
(Somehow I had initially read it as being more strict, but it is the opposite.)

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -3969,6 +3976,7 @@ PyArray_FromIter(PyObject *obj, PyArray_Descr *dtype, npy_intp count)
             ((PyArrayObject_fields *)ret)->data = new_data;
         }
         PyArray_DIMS(ret)[0] = i + 1;
+        PyArray_STRIDES(ret)[0] = elsize;
```
When exactly does this matter?

seberg replied:
FromIter is a bit weird here (I have a PR that cleans it up a bit), but I am also confused that it should matter, since the array should be empty to begin with.

The problem is that `FromIter` sometimes resizes the array (i.e. grows it dynamically), and in that case the stride has to be sane.  (This only matters if `__length_hint__ == 0` probably)

seberg replied:
The PR is: gh-20993

mattip replied:
> This only matters if `__length_hint__ == 0`

Exactly the case that tripped this: `np.fromiter((x**2 for x in range(24)), int)`. So `__length_hint__` ==0, and the code creates a 1d array with shape `(0)` and strides (which before this PR was `elsize=8`) `(0)`. Then the code calls `next()` to get values from the iterator, expanding the result for each valid  value. Since the strides is `(0)`, the end result is a 1d array with shape `(n)` but strides `(0)`, which is wrong.

eric-wieser replied:
I think this would make more sense on line 3966 when the array is reallocated, rather than doing it on every iteration.

mattip replied:
Rebased this on top of #20993, and moved the stride fixup to a better location (where we are sure `ret.size > 0`

seberg commented:
Thanks for restoring the max-size thing!

Yeah, `RELAXED_STRIDES` should have caught most of such issues.  In this case, the array is empty, so using its stride is pretty unlikely.  In practice, I expect it only affects thingse if you later grow the array and assume the strides are still compatible.
Growing an array from empty to non-empty is hopefully exceedingly rare or non-existing outside of NumPy proper.

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -4022,6 +4021,14 @@ PyArray_FromIter(PyObject *obj, PyArray_Descr *dtype, npy_intp count)
         }
         ((PyArrayObject_fields *)ret)->data = new_data;
+
+        if (count == -1) {
```
Possibly safer just to this unconditionally, in case the `count` logic changes later. It might even be faster to an irrelevant degree.

mattip replied:
Done

seberg replied:
Just to note, I kept this (added an `|| RELAXED_STRIDES_DEBUG` because I thought it might be necessary, but I guess it probably is not; remove again?).

The reason is, that the actual resetting needs to reset all strides in case of subarrays, so the whole thing is a bit more complicated.

mattip committed:
from review, unconditionally set the stride

seberg commented:
@mattip I just realized, I had that `PyArray_NBYTES_ALLOCATED` helper for the reduction code.  I am a bit surprised the reduce tests don't print the mismatch stuff now due to that not also being updated?
Not worried though, but it would still be good to update that inline-function.

seberg committed:
BUG: Fix FromIter stride resetting logic

We have to reset all strides in the case of a subarray dtype which
may have caused the result array to not be 1-D.  Since the logic
is a bit complicated now, restore the logic to only do this when it
really is necessary.

seberg committed:
MAINT: Remove now unnecessary `PyArray_NBYTES_ALLOCATED` helper

seberg commented:
Ah, this is probably an invalid access in our sum implementation.  The reduce path would be taken here with `n == 0` (unless we prevent that earlier, and I do not think we do):
https://github.com/numpy/numpy/blob/62293d45fb31f81bc7df8038a6d1c982ac05b010/numpy/core/src/umath/loops_arithm_fp.dispatch.c.src#L523-L532

That path clearly always loads one data-point from the output (even for `n == 0`), and will do an operation for said data point.  Even if that operation doesn't generate a NaN (which seems unlikely), that would give a warning if the loaded garbage happened to be a signalling NaN.

seberg commented:
So for the windows thing, I think the correct fix (which is a bug-fix in any case), is to never do an empty reduce-loop.

There are two ways to do that:
1. Modify the `IS_BINARY_REDUCE` macro to reject `dimensions == 0` (or `dimensions <= 1`)
2. Guarantee that we never call a ufunc loop on 0 elements to begin with?

Since I do not think complex iteration paths can generate length 0 inner-loop calls, both seem very feasible to me, and I think I will make a PR for it.  But I would also be happy with the second approach.

mattip commented:
I think this PR is growing too large. The change to prevent ufuncs looping when `PyArray_SIZE() == 0` should be merged first, then we can come back to this PR.

mattip committed:
ENH: short-cut trivial ufunc loops when array.size == 0

mattip commented:
After a one-line check for 0-length loops in `try_trivial_single_output_loop`, the PR is passing. Maybe more is needed, would a valgrind check catch illegal indexing into 0-length arrays? If we still have doubts, it may be safer to allocate `elemsize` bytes rather than 1 byte.

seberg commented:
If we fix it in the ufunc setup code, I think we should add an assert to `IS_BINARY_REDUCE` (which also documents the guarantee a bit).  Otherwise, I don't think more is needed as such.  I was trying to think of a good tests (and maybe have a quick look at accumulate).
Did you want to split out that change, anyway (Or I can do)?

mattip commented:
> I think we should add an assert to `IS_BINARY_REDUCE`

Is there precedence for this in the code? The macros are used like `if (IS_BINARY_REDUCE())` so I am not sure how to add an `assert` to the macro itself

> Did you want to split out that change, anyway (Or I can do)?

I originally did, but it turned out to be simpler than I thought, so I am not sure it is necessary as a separate PR.

seberg commented:
> Is there precedence for this in the code? The macros are used like if (IS_BINARY_REDUCE()) so I am not sure how to add an assert to the macro itself

Ah, I had not thought about it, but this should work (just tested briefly and seems good):
```
#define IS_BINARY_REDUCE (assert(*dimensions != 0),  \
        (args[0] == args[2])  \
        && (steps[0] == steps[2])  \
        && (steps[0] == 0))
```
Maybe a comment makes sense, that it uses the `(expr, result)` style to get to `result` only.

---

One thing here is that we still need to figure out the `arr.dtype = ...` code, that might get a bit messy.  But lets just keep it for now, and I might add test in a followup.

mattip commented:
#21499 was merged, Is there any outstanding review for this PR (pushing on it a bit since it removes the annoying "uh oh" output when testing allocation policy changes)?

seberg commented:
@mattip unfortunatley, I don't trust the `arr.dtype = ...` assignment code at all, this is the issue that the current version runs into, I believe:
```python
arr = np.ones(1, dtype="i8")
arr = np.broadcast_to(arr, 10)
arr.dtype = "i1"
print(arr)  # oops
```
Unfortunately, I don't have a clear idea of how to fix that.  Right now, I mainly see the option of making a "deep clean" here and rethinking the code almost as a whole, and I didn't try to think it through yet.
(I had briefly enterntained whether allocating a single element could help, but I don't think so.)

mattip committed:
BUG: ensure view of non-empty array with strides == (0,) raises

mattip commented:
I pushed a fix which will allow a view of an empty array (`np.prod(arr.shape) == 0`), but restores the failure to set a dtype on a non-empty array when the strides do not match the `elemsize`. The check now requires `np.prod`, is that too expensive?

seberg commented on numpy/core/src/multiarray/getset.c:
```diff
@@ -513,6 +516,7 @@ array_descr_set(PyArrayObject *self, PyObject *arg, void *NPY_UNUSED(ignored))
         /* resize on last axis only */
         int axis = PyArray_NDIM(self) - 1;
         if (PyArray_DIMS(self)[axis] != 1 &&
+                PyArray_SIZE(self) != 0 &&
```
Hah, this should work, yeah.  So the one thing is that if the stride was previously 0, in theory it should remain 0 in this case (but gets replaced with `elsize`).  I think I am OK with ignoring the potential UB problem that may follow that for now.

If we really want to fight it, I guess we would have to add asserts into `array_dealloc` to flush out places that need adjustment.

@eric-wieser how worried are you actually about the UB?  If not very, I am tempted to suggest accepting that this path might lead to this UB.

seberg replied:
@mattip if you really care about the prod, you cold order it to the end, since it will be hit very rarely then?
But honestly, this is still a Python "call", so I really doubt it matters at all.

seberg commented:
OK, I am going to put this in a bit optimistically, since I think Matti was hoping to get this into the next release.  If we got holes in this logic, it seems OK to fix it in bug-fix releases as well.

But if Eric (or anyone else) is a bit worried about the changes, we can revert again as well.   However, right now, I also think that downstream testing might be good here, and merging will get us at least a bit of testing.

mattip commented:
:rocket:
